### ISSUE TITLE
[rfc] how best to reduce runc's effects on cgroup memory limits

### ISSUE THREAD

**Issue by cyphar** [opencontainers/runc#4021]
> Temporarily "unresolving" to un-hide on GitHub, and maybe there's some follow-up discussions to be had (NOT for this PR fwiw) but we can move the comment elsewhere.
>
> We initially set this limit to 4MB, and raised it to 6MB when later versions of runc became more memory-hungry (I think that was part of a security fix?)
>
> While looking that up, I found;
>
> - our initial minimum was 512K (back in the LXC days); https://github.com/moby/moby/commit/9ee11161bf0c635a0b0164ae1dee806c6fcfe7b7 (https://github.com/moby/moby/pull/892)
> - it was updated to 4MB (after we switched to libcontainer); https://github.com/moby/moby/commit/9b755412ab03816f0b7d6013618c65ae68781bcb (https://github.com/moby/moby/pull/8173)
>     - https://github.com/moby/moby/pull/8173#issuecomment-56442803 "This is because over time the docker binary has grown and because we are linking to more libraries it takes more memory to load the docker binary as the container's init process." 
> - it was raised to 6MB to account for runc requiring more in https://github.com/moby/moby/commit/d2e23405bed2f5388b4a6c4f50c94ed2ab580319 (https://github.com/moby/moby/pull/41168)
>
> Honestly, from Moby/Docker's (and other "higher level" runtimes, I don't think it's "ideal"
>
> - It requires higher-level runtimes to have knowledge about the implementation's requirements
> - It requires them to (hard-code) based on the "worst case" scenario (e.g. would `crun` have a lower requirement?)
> - It limits users to set the limit to what _their container process_ requires _at runtime_ (which may be _lower_ than memory needed during init).
>
> This is orthogonally related to a (rather heated) discussion we had recently about https://github.com/moby/moby/pull/45511, during which a question was raised _"what does the runtime-spec actually describe?"_
>
> - Is it declarative? Does it define the container "at runtime" (after `execve`) to execute, meaning: the spec is "declarative", and defines the user's intent w.r.t. describing requirements for the container(s process).
> - Or does it describe the container _including_ the initial `init` (`runc`) to execute: the spec defines what's needed to construct the container (including capabilities and resources needed to do so).
>
> The difference is very subtle, and in _MOST_ cases, they are interchangeable (which is likely why they were never explicitly defined), however there are situations such as the one being discussed here (and the one I linked), where 
>
> My (personal) ideal would be to somehow get to the "declarative" variant; the user describes their intent ("make it so!"), and the runtime(s) take care of "whatever is needed" to make it happen. This would be the most user-friendly approach, and most "portable", as it would remove implementation-specific conditions out of the equation. But of course, this may not be something that can be realized (easily), beyond that ship possibly having sailed (runc effectively being the "reference implementation", and the runtime-spec usually being adjusted to fit (not the reverse)).
>
> That said, _MAYBE_ there are some options if we take the "features" option into account (https://github.com/opencontainers/runtime-spec/pull/1130); perhaps an addition there could _either_ specify the minimum requirements for the runtime that's used, or advertise the runtime's own requirements (i.e.; for this specific case: add XXk of memory to what's specified in the provided OCI spec).

_Originally posted by @thaJeztah in https://github.com/opencontainers/runc/pull/3987#discussion_r1327940441_

**Comment by cyphar** [comment]
> To be honest I don't think we can have a hard guarantee for memory usage, peak memory usage is not deterministic for one (and, as evidenced by the CentOS test failures it seems that sometimes we can end up using 2.5mb more in some environments).
> 
> I think we also can't do something simple like applying the cgroup setting later, and we have to join the cgroup to unshare the cgroup namespace.
>
> Maybe we could work around it by joining the cgroup just for cgroupns creation, leave the cgroup, and then get re-attached right before the exec. It would make things more complicated but it would take runc out of the equation for memory usage. We couldn't do this for containers that have `CAP_SYS_PTRACE` (for the same reasons as with runc-dmz) because an attacker could inject code into the joining process and run outside of cgroup restrictions. For `runc exec` we could just delay cgroup attach because we don't need to create the namespace. @kolyshkin is this idea insane?

_Originally posted by @cyphar in https://github.com/opencontainers/runc/pull/3987#discussion_r1328041488_

**Comment by thaJeztah** [comment]
Thanks! Also /cc @neersighted who was in one of those conversations at the time.

**Comment by cyphar** [comment]
@rata That is very similar to how runc currently works. We fork a child (`runc init`) and we give it a parsed form of the container configuration. We have to do an exec because of how Go works (there's no way around that, sadly) but the idea is very similar.

**Comment by rata** [comment]
@cyphar no, it is very different because, while runc forks several times, it never execs into a very small binary to apply the cgroup limits. It just apply them while running the runc binary. Therefore, the runc binary of ~13MB imposes a limit for the cgroup mem limit.

**Comment by cyphar** [comment]
The runc binary size doesn't affect the cgroup limit. I can run a 2.5mb limit container (with `RUNC_DMZ=legacy` no less) on my machine. If the runc binary size had any effect on memory usage that shouldn't be possible.

I think this is because we never re-exec runc after joining the cgroup so the memory accounting for the binary is attributed to the host. The memory usage also isn't from binary cloning, I think that has been pretty well-established at this point.

My pet theory was that runc exits so quickly that the GC doesn't have a chance to run (if you try to run with `GODEBUG=gctrace=1`, there is no output) and so we are just looking at the total memory usage of runc without anything being freed. But some playing around with `debug.FreeOSMemory` seems to indicate that we might actually be using all of that memory.
